docs: document AI assistance in the ES|QL editor (#6552, #6681) (#6714)
elastic/docs-content: explore-analyze/query-filter/languages/esql-kibana.md

| {kbd}`cmd+/`       | {kbd}`ctrl+/`       | Comment or uncomment the current line or selected lines |
| {kbd}`cmd+i`       | {kbd}`ctrl+i`       | [Prettify query](#_make_your_query_readable) {applies_to}`stack: ga 9.4+` |
| {kbd}`cmd+j`       | {kbd}`ctrl+j`       | [Generate {{esql}} from a `//` comment](#esql-kibana-ai-comment) {applies_to}`stack: preview 9.5+` |
| {kbd}`cmd+k`       | {kbd}`ctrl+k`       | Open the [search bar](#esql-kibana-quick-search) |

:::{tip}
You can find the list of shortcuts directly from the editor. Look for the ![keyboard](../../images/keyboard.svg "keyboard =2%") icon.
:::

### Build ES|QL queries from KQL syntax [esql-kibana-quick-search]
```{applies_to}
serverless: preview
stack: preview 9.3+
```

The {{esql}} editor includes a search bar that helps you build a query without writing the full {{esql}} syntax. To open it, select the {icon}`magnify` or {icon}`magnify_sparkles` search icon in the editor's toolbar, or press {kbd}`cmd+k` (Mac) or {kbd}`ctrl+k` (Windows/Linux). The search bar offers two modes:

- **KQL**: filter your data with free text or [KQL](kql.md) syntax, as described in this section.

> - {applies_to}`stack: preview 9.5+` {applies_to}`serverless: preview` **Natural language**: describe the query you want in plain language and let an LLM generate it for you. Refer to [Generate a full query from natural language](#esql-kibana-quick-search-nl).

Reviewer: ```suggestion
- {applies_to}`stack: preview 9.5+` {applies_to}`serverless: preview` **Natural language**: describe the query you want in plain language and let an AI agent generate it for you. Refer to [Generate a full query from natural language](#esql-kibana-quick-search-nl).
```

I realized in a later edit that this might be more precise than simply "LLM"

The search bar closes automatically when you start typing in the editor or select outside of it.

To filter your data with KQL:

1. Open the editor's search bar ({icon}`magnify` or {icon}`magnify_sparkles`).
2. Select the data sources to search.
3. Type the text you want to search for as free text or using [KQL](kql.md) syntax.
4. Submit your search by pressing {kbd}`enter`. The generated query includes a `FROM` command based on the data sources you selected (or `TS` if the data source is a time series data stream), and a `WHERE KQL()` command that contains the text you typed in the search bar. The query is saved to your [query history](#esql-kibana-query-history) so you can restore it later.
5. Refine your query with any other {{esql}} command or function that you need.

![Search bar in the ES|QL editor](https://images.contentstack.io/v3/assets/bltefdd0b53724fa2ce/bltc3b8614d0ecabbd9/69ebb647065c54efe579b251/esql-quick-search-kql.gif "=60%")

### Write and fix queries with AI [esql-kibana-ai-assistance]
```{applies_to}
stack: preview 9.5+
serverless: preview
```

> When your deployment or project has a configured large language model (LLM) connector, {{kib}} can use AI to help you author {{esql}}. The editor uses the same default AI connector as {{kib}}'s other AI features, such as {{agent-builder}}. It isn't a connection that you configure for the editor specifically.

Reviewer: ```suggestion
When your deployment or project has a configured large language model (LLM) connector, {{kib}} can use AI to help you author {{esql}}. The editor uses the same default AI connector as {{kib}}'s other AI features, such as {{agent-builder}}. You don't need to configure a specific connection for the editor.
```

nit: active voice

Choose the entry point that matches what you want to do:

- [**Generate a full query from natural language**](#esql-kibana-quick-search-nl) in the editor's search bar, when you're starting from scratch or want to replace the whole query.
- [**Generate {{esql}} from a comment**](#esql-kibana-ai-comment) to add or change a single pipe inside the query you're writing.
- [**Fix query errors with AI**](#esql-kibana-ai-fix) when a query fails validation.

**Requirements**

- For {{ech}}, {{ece}}, and {{eck}} deployments or self-managed clusters, you need an Enterprise license.
- A configured LLM connector. Refer to [Configure access to LLMs](/explore-analyze/ai-features/llm-guides/llm-connectors.md).

> Without these requirements, the AI prompts and actions don't appear and the editor uses only its standard autocomplete behavior. In the search bar, if no connector is available, you're prompted to set one up.

Reviewer: ```suggestion
Without these requirements, the AI prompts and actions don't appear and the editor uses only its standard autocomplete behavior. If no connector is available in the search bar, you're prompted to set one up.
```

#### Generate a full query from natural language [esql-kibana-quick-search-nl]

> You can describe the query you want in plain language and let an LLM translate it into {{esql}}. This is useful when you know what you want to ask of your data but are not sure which {{esql}} commands or functions to use.

Reviewer: ```suggestion
Not sure how to write an {{esql}} query? Describe what you want from your data in natural language and a specialized AI agent writes the query for you.
```

nits: might be a bit more concise and to the point 

uses "natural language" for consistency and "specialized AI agent" is a bit more precise than "LLM"

> You can describe the query you want in plain language and let an LLM translate it into {{esql}}. This is useful when you know what you want to ask of your data but are not sure which {{esql}} commands or functions to use.

Reviewer 2: ```suggestion
You can describe the query you want in plain language and let an LLM translate it into {{esql}}. Use this when you have a specific question about your data, but are not sure which {{esql}} commands or functions to use.
```

1. Open the editor's search bar ({icon}`magnify_sparkles`).
2. From the mode selector, select **Natural language**.
3. In the input, describe the query you want. For example, `Show the average response time per host for the last 24 hours`.

> 4. Submit your request by pressing **Enter**. The editor replaces or updates the current query, runs it, and saves it to your [query history](#esql-kibana-query-history).

Reviewer 2: ```suggestion
4. Press **Enter** to submit your request. The editor replaces or updates the current query, runs it, and saves it to your [query history](#esql-kibana-query-history).
```

5. Review the generated query and refine it with any other {{esql}} command or function that you need.

:::{tip}
The current query in the editor is sent to the LLM as context, so you can ask follow-up requests that build on it. For example, after running a generated query, ask `Group the results by region` to extend it.
:::

:::{image} /explore-analyze/images/kibana-esql-search-bar-nl.png
:alt: Natural language mode in the ES|QL editor search bar
:screenshot:
:width: 60%
:::

#### Generate {{esql}} from a comment [esql-kibana-ai-comment]

> Write a `//` comment that describes what you want, then press {kbd}`cmd+J` (Mac) or {kbd}`ctrl+J` (Windows/Linux). The editor sends your comment to the LLM and inserts the generated {{esql}} on the next line.

Reviewer: ```suggestion
Write a `//` comment that describes what you want, then press {kbd}`cmd+J` (Mac) or {kbd}`ctrl+J` (Windows/Linux). The editor sends your comment to the AI agent and inserts the generated {{esql}} on the next line.
```

You can use this to:

> - Start a query from scratch when the editor is empty. For example, type `// Show the top 10 destinations by flight count` and press {kbd}`cmd+J` to generate a complete query.

Reviewer: ```suggestion
- Write a query from scratch when the editor is empty. For example, type `// Show the top 10 destinations by flight count` and press {kbd}`cmd+J` to generate a complete query.
```

> - Add a step to an existing query. Place the comment between pipes (for example `// filter for delayed flights`) and the LLM generates a single pipe to append. If the LLM determines that your comment describes a change to the next pipe rather than a new step, it modifies that pipe instead.

Reviewer: ```suggestion
- Add a step to an existing query. Place the comment between pipes (for example `// filter for delayed flights`) and the AI agent generates a single pipe to append. If the AI agent determines that your comment describes a change to the next pipe rather than a new step, it modifies that pipe instead.
```

feel free to ignore these if disagree, just making them suggestions so can commit quickly if want to :)

As you write, the editor displays inline hints that remind you how to use the feature: on an empty line it suggests starting a `//` comment, and once you've written one it prompts you to press the shortcut. The placeholder in an empty editor conveys the same order: describe what you want in a `//` comment first, then press the shortcut to generate {{esql}} from it.

> While the LLM is generating, an italic `Generating...` indicator appears next to your comment. Pressing the shortcut again cancels the in-flight request and starts a new one.

Reviewer: ```suggestion
While the AI agent is generating, an italic `Generating...` indicator appears next to your comment. Pressing the shortcut again cancels the in-flight request and starts a new one.
```

> While the LLM is generating, an italic `Generating...` indicator appears next to your comment. Pressing the shortcut again cancels the in-flight request and starts a new one.

Reviewer 2: ```suggestion
While the LLM is generating, an italic `Generating...` indicator appears next to your comment. Press the shortcut again to cancel the in-flight request and start a new one.
```
Is the mention of "italic" relevant?

Author: ```suggestion
While the AI agent is generating, a `Generating...` indicator appears next to your comment. Press the shortcut again to cancel the in-flight request and start a new one.
```

When the generated code is ready, the editor highlights it and shows review actions:

- **Keep** ({kbd}`cmd+shift+enter` on Mac, {kbd}`ctrl+shift+enter` on Windows/Linux) accepts the change and leaves your comment in place so you can refine it and regenerate.
- **Undo** ({kbd}`cmd+shift+backspace` on Mac, {kbd}`ctrl+shift+backspace` on Windows/Linux) removes the generated code.

:::{image} /explore-analyze/images/kibana-esql-generate-from-comment.png
:alt: The ES|QL editor showing a query generated from a comment, highlighted with Keep and Undo review actions
:screenshot:
:width: 50%
:::

> When the LLM rewrites an existing pipe instead of adding one, the original pipe appears with a strikethrough and the **Keep** button becomes **Replace**, indicating that accepting the suggestion removes the original pipe.

Reviewer: ```suggestion
When the AI agent rewrites an existing pipe instead of adding one, the original pipe appears with a strikethrough and the **Keep** button becomes **Replace**, indicating that accepting the suggestion removes the original pipe.
```

#### Fix query errors with AI [esql-kibana-ai-fix]

> When your query fails validation, hover over the underlined error in the editor. The error popup includes a **✨ Fix with AI** link. Select it to send the query and the error to the LLM and have it propose a corrected version.

Reviewer: ```suggestion
When your query fails validation, hover over the underlined error in the editor. The error popup includes a **✨ Fix with AI** link. Select it to send the query and the error to the AI agent and have it propose a corrected version.
```
